sshd-core/src/main/java/org/apache/sshd/common/session/helpers/AbstractSession.java

```diff
@@ -109,6 +109,16 @@
      */
     public static final String SESSION = "org.apache.sshd.session";
 
+    /**
+     * A last-resort timeout for waiting after having received a KEX_INIT message from the peer until we have prepared
+     * our own KEX proposal. This timeout should actually never be hit unless there is a serious deadlock somewhere and
+     * the session is never closed.
+     *
+     * @see <a href="https://issues.apache.org/jira/browse/SSHD-1197">SSHD-1197</a>
+     * @see #doKexNegotiation()
+     */
+    private static final Duration KEX_PROPOSAL_SETUP_TIMEOUT = Duration.ofSeconds(42);
```

Comment:
Please use a `CoreModuleProperties#KEX_PROPOSAL_SETUP_TIMEOUT` as initially suggested so that users can decide to shorten/lengthen this value instead of it being hardcoded

Reply:
Hmmm. If we make that configurable, users can very easily shoot themselves in the foot by setting a (very) low value, like 0. I don't like that. Lengthening this should never be necessary. The thread that is preparing this side's proposal is already running.

Reply:
>>>  If we make that configurable, users can very easily shoot themselves in the foot by setting a (very) low value, like 0.

I understand your concern, but personally I believe in total user responsibility - I don't like being told by someone that "they know better", so I am reluctant to impose my opinion on others (BTW, this is also why I always insist that all our internal variables and/or methods be accessible either via getters or as public/protected members).

In this context, the same could be said for *all* our configuration values - users can shoot themselves in the foot by choosing a wrong or conflicting value. However, I believe this is a good thing - they will learn... IMO, we are not supposed to protect our users from mistakes that occur because they do not bother to read the documentation or understand the consequences of their actions. In this case, perhaps we could do something that would satisfy both our concerns:
```java
Duration waitTime = CoreModuleProperties.KEX_PROPOSAL_SETUP_TIMEOUT.getRequired(this);
ValidateUtils.checkTrue(waitTime.toMillis() > 0L, "Infinite/Undefined value N/A: %d", waitTime);
```

>> Lengthening this should never be necessary. The thread that is preparing this side's proposal is already running.

While it seems reasonable, we don't know what weird usages our users might encounter or what unforeseen circumstances. Therefore, I believe we should give them the flexibility to decide for themselves. 

Bottom line - I trust your judgement, and if you still feel that a constant value is more appropriate I will not object to it. In any case, let's also update the README (perhaps add a KEX section) and explain this mechanism so that if there are future questions or suspicions of bugs around this we will have a clear picture of the implementation and the choice of either a constant or configurable timeout value.

Reply:
I've been thinking more about this -- I'll add it with validation for now. Reason: we are indeed waiting until sendKexInit has sent the buffer (or rather, has queued it for sending on the IOSession). But I'm still not happy with that whole KEX implementation. In particular, this having to wait until the buffer is sent should not be necessary, but with the current implementation is unavoidable because of the `ReservedSessionMessagesHandler.sendKexInitRequest` hook added for SSHD-1097, which can modify the buffer and even might send it on its own.

That hook was added recently, and is used only in that "endless tarpit" thing, where it actually only suppresses writing the buffer. If that were changed to something that could only say "don't send the message", I could extract that whole KEX state machine in a separate class and it would even possible to do an implementation that didn't have to wait at all for this race condition case handled here.

I wouldn't add a full explanation of how KEX works, that's explained in the RFCs. But the new timeout property will have a comment explaining what it is for.

Reply:
Sounds reasonable to me